fsspec/spec.py

```diff
@@ -115,6 +119,87 @@ class AbstractFileSystem(metaclass=_Cached):
     #: Extra *class attributes* that should be considered when hashing.
     _extra_tokenize_attributes = ()
 
+    # Set by _Cached metaclass
+    storage_args: Tuple[Any, ...]
+    storage_options: Dict[str, Any]
+
+    class _JSONEncoder(json.JSONEncoder):
+        def default(self, o: Any) -> Any:
+            if isinstance(o, AbstractFileSystem):
+                return o.to_dict()
+            if isinstance(o, PurePath):
+                cls = type(o)
+                return {"cls": f"{cls.__module__}.{cls.__name__}", "str": str(o)}
+
+            return super().default(o)
+
+    class _JSONDecoder(json.JSONDecoder):
+        def __init__(
+            self,
+            *,
+            object_hook: Optional[Callable[[Dict[str, Any]], Any]] = None,
+            parse_float: Optional[Callable[[str], Any]] = None,
+            parse_int: Optional[Callable[[str], Any]] = None,
+            parse_constant: Optional[Callable[[str], Any]] = None,
+            strict: bool = True,
+            object_pairs_hook: Optional[Callable[[List[Tuple[str, Any]]], Any]] = None,
+        ) -> None:
+            self.original_object_hook = object_hook
+
+            super().__init__(
+                object_hook=self.custom_object_hook,
+                parse_float=parse_float,
+                parse_int=parse_int,
+                parse_constant=parse_constant,
+                strict=strict,
+                object_pairs_hook=object_pairs_hook,
+            )
+
+        @classmethod
+        def try_resolve_path_cls(cls, dct: Dict[str, Any]):
+            with suppress(Exception):
+                fqp = dct["cls"]
+
+                try:
+                    path_cls = _import_class(fqp)
+                except Exception:
+                    raise
```

Comment:
why catch Exception only to raise it again, all within the outer suppress block?

Reply:
It's just to have a consistent control flow with the code for decoding the filesystem class. If you think it's too verbose then I can simplify it.